### PR TITLE
Refactor development container builds

### PR DESCRIPTION
**Pull request by portante**
We collapse the two docker files to one, using a Jinja2 template, and remove all the Red Hat internal references and RPM GPG keys (since those keys are tied to the loaded repo file).

We are not allowed to publish UBI images where we add Red Hat subscription bits on top, so we remove the method of publishing the UBI image.

### CODE REVIEW

**Review by portante — COMMENTED** [review]
Should have working tests now.

File: jenkins/development.Dockerfile.j2
Comment (dbutenhof) [inline jenkins/development.Dockerfile.j2:54]: Intriguing; a comment to justify/explain this would be interesting. I assume there's a use case for this being used in our Jenkins fleet?

Comment (portante) [reply jenkins/development.Dockerfile.j2:54]: This is not used by our Jenkins fleet.  This is used by a pbench-server-in-a-can, since a UBI base doesn't provide all the RPMs and python modules we need to build it.

**Review by dbutenhof — DISMISSED** [review]

**Review by dbutenhof — DISMISSED** [review]

**Review by ndokos — DISMISSED** [review]

File: jenkins/development.Dockerfile.j2
Comment (webbnh) [reply jenkins/development.Dockerfile.j2:135]: So, I just `diff`'d the original pair of `.Dockerfile`'s, and I think I understand the conditionalization here better now.  But, my initial instinct stands:  I think we have a set of requirements which this list is supposed to express, and it should be substantially the same on both/all platforms.  We'll still want conditionalization for cases where the names are different, and we'll want conditionalization for cases where the target package is unavailable; however, I think that the number of those cases is smaller than what we now have in this PR.  I think we're seeing some "extras" here, because the Jenkins `.Dockerfile` has evolved since the UBI one was cloned from it, and the UBI one hasn't kept up.

That said, we should ponder what we want the UBI build for.  It was originally intended for the Pbench-in-a-Can image, but Vishal has pointed out, rightly, that some of our images are ~huge~quite large.  _For that purpose_, we don't need the packages for the Agent, or testing, or other utility uses...and, it might be nice if we omitted them.  On the other hand, it might be nice to have a fully-functioning UBI image for other purposes...so, I don't really have any specific recommendations on this aspect.

In any case, I would like to see less conditionalization in this file if we can swing it.

File: jenkins/run
Comment (webbnh) [inline jenkins/run:5]: While you're at it, I would get rid of the comma, too.

File: jenkins/run
Comment (webbnh) [inline jenkins/run:17]: Do we know why this script insists upon having a path which is the current working directory plus a `jenkins` subdirectory?  I _think_ it's because it wants to map the Git checkout on the host to `/src/pbench` inside the container.  If that's the case, it can do that based either on the return from `git rev-parse` or by analyzing its own path -- it shouldn't need to use `$(pwd)` for this.

Does anyone else think that we should relax the current working directory restriction, if we can?

File: server/pbenchinacan/container-build.sh
Comment (webbnh) [inline server/pbenchinacan/container-build.sh:28]: Do you need the `$(shell`, or would `$(cat` suffice?  Or, better yet, what about `$(<`?

**Review by webbnh — APPROVED** [review]
My foremost concern is whether the results of this PR will leave us with a workable process for building the UBI base image (e.g., whether we can rebuild the Pbench-in-a-Can container from scratch).  Prior to this PR, there was a bunch of knowledge captured and/or encoded into these files, and this PR is (correctly) removing that...but we as developers will need it...so where can we find it after this PR is merged?

Also, don't we need to install `epel-release-latest-8.noarch.rpm` for the UBI-based build?

Other than those details, I have the usual set of nits, suggestions, and small questions.

File: server/pbenchinacan/container-build.sh
Comment (portante) [reply server/pbenchinacan/container-build.sh:28]: Ah, right, not a Makefile ... thanks.

File: jenkins/run
Comment (portante) [reply jenkins/run:17]: This is all going to be changed in PR #2551.  So not worth addressing here.

File: jenkins/development.Dockerfile.j2
Comment (portante) [reply jenkins/development.Dockerfile.j2:183]: This was an artifact of trying to work with Fedora 35.  Just needs to be `if fedora` now.

File: jenkins/development.Dockerfile.j2
Comment (portante) [reply jenkins/development.Dockerfile.j2:135]: The goal is to have the Fedora development image have as few non-RPM installed package as possible ... ideally 0.

The UBI based image also needs the same, but we are not in that phase of productization to make a material impact on that.

I'd rather know what we have where now, rather than having to rediscover it later.

Further, we really want the development image requirements broken up a bit more:

 * common to agent and server
 * agent run-time RPMs
 * server run-time RPMs
 * composed agent & server run-time RPMs
 * agent & server test RPMs

That way, we can leverage those base images for our other containers.  But that is future work.  I'd rather keep the spirit of the previous two separate files than try to go further in this commit.

File: jenkins/development.Dockerfile.j2
Comment (portante) [reply jenkins/development.Dockerfile.j2:157]: This was a Fedora 35 problem.  Not using Fedora 35 now, so the `if False` will be removed.

File: jenkins/development.Dockerfile.j2
Comment (portante) [reply jenkins/development.Dockerfile.j2:55]: It is still worth using for Fedora.

**Comment by portante** [comment]
> My foremost concern is whether the results of this PR will leave us with a workable process for building the UBI base image (e.g., whether we can rebuild the Pbench-in-a-Can container from scratch). Prior to this PR, there was a bunch of knowledge captured and/or encoded into these files, and this PR is (correctly) removing that...but we as developers will need it...so where can we find it after this PR is merged?

I have verified that both the Fedora and UBI image builds work, the fedora image is pushed to quay.io and used in the tests.

The UBI image is not pushed anywhere, but I posted the internal references in our chat channel for now.  We can move them to a better place that we decide as an internal team.

> 
> Also, don't we need to install `epel-release-latest-8.noarch.rpm` for the UBI-based build?

No because that is part of the internal `build.repo` we provide via the environment variable.

> 
> Other than those details, I have the usual set of nits, suggestions, and small questions.

Thanks for the review!

I'll be posting an update addressing the comments shortly.

**Review by portante — COMMENTED** [review]
Ready for another round of reviews!

**Review by dbutenhof — APPROVED** [review]

File: jenkins/development.Dockerfile.j2
Comment (webbnh) [reply jenkins/development.Dockerfile.j2:137]: I concur with all of that...I just want it to be clear why each of the conditionalized packages is conditionalized.

Taking `python3-bottle` for instance.  That package is required for the Agent's tool data sync.  So, the only reason that it should be conditionalized is if it is not available as an RPM.  As it happens, the RPM _is_ available on Fedora but (apparently) _not_ for the UBI.  So, I'm thinking that the conditionalization should be excluding UBI rather than selecting for Fedora.

That is, adopting the perspective that this recipe will be used for more than just two platforms (e.g., what about Centos; and what about RHEL 9 and Fedora Rawhide), I think we should exercise some care in how we construct the criteria.

As I said above, our requirements aren't dependent upon the platform (they are driven by the code), so at most we should be _excluding_ things here, not conditionally **_including_** things.

**Review by webbnh — DISMISSED** [review]
So...other than the fact that I cannot build the UBI image and that I think we should use a different approach to conditionalization in the Dockerfile, I'm OK with this PR.   🤪

File: jenkins/development.Dockerfile.j2
Comment (portante) [reply jenkins/development.Dockerfile.j2:137]: Sure, I can invert all the conditionals to be `not ubi`.

File: jenkins/development.Dockerfile.j2
Comment (webbnh) [reply jenkins/development.Dockerfile.j2:49]: I was able to build successfully from another VM.

**Review by portante — COMMENTED** [review]
Inverted the template conditionals to exclude only `ubi`.

Rebuilt both the Fedora and UBI images, pushed the Fedora image (now tagged with `main`) to quay.io, and the CI jobs should be using the final bits.

**Review by dbutenhof — DISMISSED** [review]

**Review by portante — COMMENTED** [review]
Okay, ready for another review after removing as many conditionals as possible.

**Review by webbnh — APPROVED** [review]
GTG

**Review by dbutenhof — APPROVED** [review]

**Comment by dbutenhof** [comment]
The ball, Pete, is in your court; I believe this can go in despite the stale "DO NOT MERGE" (because you removed your "fake" branch several pushes ago), but I'll leave that to you...

**Pull request merged**
